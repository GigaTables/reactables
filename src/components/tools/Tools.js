import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Button from '../form/Button';
import PagesSelector from './PagesSelector';
import Search from './Search';
import CSVLink from "../form/CSVLink";
import styles from '../../css/styles.css';

let CommonConstants = require('../CommonConstants');
let EditorConstants = require('../EditorConstants');
let Lang = require('../Lang');

class Tools extends Component {
    getPagesSelection() {
        const {
            struct,
            display,
            defaultPerPage,
            perPage,
            updatePerPage,
            perPageRows,
            lang,
        } = this.props;

        if (struct.rowsSelector.indexOf(display) === -1) {
            return '';
        }
        return (<PagesSelector
            perPage={perPage}
            updatePerPage={updatePerPage}
            defaultPerPage={defaultPerPage}
            perPageRows={perPageRows}
            lang={lang}/>)
    }

    getSearch() {
        const {
            struct,
            display,
            search,
            doSearch,
            lang,
        } = this.props;

        if (struct.search.indexOf(display) === -1) {
            return '';
        }
        return (<Search
            search={search}
            doSearch={doSearch}
            lang={lang}/>)
    }

    getButtons() {
        const {
            selectedRows,
            showPopup,
            lang,
            tableOpts,
            display,
            jsonData,
            struct,
        } = this.props;

        let language = Lang[lang];
        let buttons = [];
        if (typeof tableOpts.buttons !== CommonConstants.UNDEFINED
            && tableOpts.buttons.length > 0
            && tableOpts.buttonsPosition.indexOf(display) !== -1) {
            tableOpts.buttons.map((btn, i) => {
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_CSV
                    && jsonData !== CommonConstants.UNDEFINED
                    && typeof struct.download !== CommonConstants.UNDEFINED && struct.download.csv === true) {
                    buttons[i] = <CSVLink
                        active={false}
                        jsonData={jsonData}
                        key={i}>CSV Export</CSVLink>;
                }
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_CREATE) {
                    buttons[i] = <Button
                        active={false}
                        action={EditorConstants.ACTION_CREATE}
                        showPopup={showPopup}
                        key={i}>{language.editor_create}</Button>;
                }
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_EDIT) {
                    buttons[i] = <Button
                        active={!(selectedRows.length >= 1)}
                        selectedRows={selectedRows}
                        action={EditorConstants.ACTION_EDIT}
                        showPopup={showPopup}
                        key={i}>{language.editor_edit}</Button>;
                }
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_REMOVE) {
                    buttons[i] = <Button
                        active={(selectedRows.length === 0)}
                        selectedRows={selectedRows}
                        action={EditorConstants.ACTION_DELETE}
                        showPopup={showPopup}
                        key={i}>{language.editor_remove}</Button>;
                }
            });
        }
        return buttons;
    }

    render() {
        const {
            selectedRows,
            showPopup,
            lang,
            tableOpts,
        } = this.props;

        let language = Lang[lang];
        let buttons = [];
        if (typeof tableOpts.buttons !== CommonConstants.UNDEFINED) {
            tableOpts.buttons.map((btn, i) => {
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_CREATE) {
                    buttons[i] = <Button
                        active={false}
                        action={EditorConstants.ACTION_CREATE}
                        showPopup={showPopup}
                        key={i}>{language.editor_create}</Button>;
                }
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_EDIT) {
                    buttons[i] = <Button
                        active={(selectedRows.length === 1) ? false : true}
                        selectedRows={selectedRows}
                        action={EditorConstants.ACTION_EDIT}
                        showPopup={showPopup}
                        key={i}>{language.editor_edit}</Button>;
                }
                if (btn[EditorConstants.EXTENDED] === EditorConstants.EDITOR_REMOVE) {
                    buttons[i] = <Button
                        active={(selectedRows.length === 0) ? true : false}
                        selectedRows={selectedRows}
                        action={EditorConstants.ACTION_DELETE}
                        showPopup={showPopup}
                        key={i}>{language.editor_remove}</Button>;
                }
            });
        }
        return (
            <div className="gt_head_tools">
                {this.getButtons()}
                {this.getPagesSelection()}
                {this.getSearch()}
                <div className={styles.clear}></div>
            </div>
        )
    }
}

Tools.propTypes = {
    updatePerPage: PropTypes.func.isRequired,
    perPageRows: PropTypes.array,
    doSearch: PropTypes.func,
    tableOpts: PropTypes.object,
    showPopup: PropTypes.func,
    defaultPerPage: PropTypes.number,
    perPage: PropTypes.number,
    search: PropTypes.string,
    selectedRows: PropTypes.array,
    lang: PropTypes.string,
};

export default Tools
