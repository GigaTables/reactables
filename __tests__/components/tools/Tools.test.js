import React from 'react'
import renderer from 'react-test-renderer'
import Tools from '../../../src/components/tools/Tools.js'

const editor = {
    ajax: 'http://gigatables.loc/editor.php',
    ajaxFiles: 'uploadFiles.php',
    struct: {
        buttons: [ 'top', 'bottom' ] // buttons
    },
    fields: [
        {
            label: 'ID',
            name: 'id',
            type: 'hidden'
        },
        {// an example of using select - automatically selected if matches with data in table column
            label: 'Types:',
            name: 'types[]',
            values: [// if select,checkbox,radio etc types - need this pre-set structure of values
                { 'key1': 'val1' },
                { 'key2': 'val2' }
            ],
            type: 'checkbox' // select,checkbox,radio
//              attrs: [
//                {'multiple':true}
//              ]
        },
        {
            label: 'Article title:',
            name: 'title',
            type: 'text', // default, other: password, file, select, multiselect etc
            attrs: [
                { 'pattern': '^[A-Za-z0-9_]+$' }
            ]
        },
        {
            label: 'Description:',
            name: 'desc',
            type: 'textarea'
        },
        {
            label: 'Date Time:',
            name: 'date',
            type: 'date'
        },
        {
            label: 'Image:',
            name: 'image',
            type: 'file'
        }
    ]
}

it('renders Tools correctly', () => {
    const tree = renderer.create(
        <Tools
            defaultPerPage={100}
            display="bottom"
            lang="en"
            perPage={50}
            perPageRows={[ 25, 50, 100, 200 ]}
            search=""
            selectedRows={[]}
            struct={{
                search: [ 'top', 'bottom' ],
                rowsSelector: [ 'desc', 'top', 'bottom' ],
                pagination: [ 'bottom' ],
                download: {
                    csv: true
                },
            }}
            tableOpts={{
                buttons: [
                    { // reload button to fetch content manually
                        extended: 'editor_reload',
                        editor: editor,
                        triggerAfter: (() => {
                            console.log('after reload')
                        }), triggerBefore: (() => {
                            console.log('before reload')
                        })
                    },
                    {// to set CSV Download button
                        extended: 'editor_csv',
                        editor: editor
                    },
                    {
                        extended: 'editor_create', editor: editor, triggerAfter: (function () {
                            console.log('after create')
                        }), triggerBefore: (function () {
                            console.log('before create')
                        })
                    },
                    {
                        extended: 'editor_edit', editor: editor, triggerBefore: (function () {
                            console.log('before edit')
                        })
                    },
                    {
                        extended: 'editor_remove', editor: editor, triggerAfter: (function () {
                            console.log('after del')
                        })
                    }
                ],
                buttonsPosition: [ 'top', 'bottom' ],
                theme: 'material-ui'
            }}
            updatePerPage={(e) => {}}
            doSearch={(e) => {}}
            jsonData={[{ "GT_RowId": 178, "title": "Test 178st row", "id": 178, "desc": "Lorem Ipsum is simply dummy 7241 text of the printing and typesetting", "info": "some info some info some info some info", "date": "22:00:22 10:05:2018", "field1": 86, "field2": 1412, "field3": 12523 },
                { "GT_RowId": 91, "title": "Test 91st row", "id": 91, "desc": "Lorem Ipsum is simply dummy 8606 text of the printing and typesetting", "info": "some info some info some info some info", "date": "22:01:49 10:05:2018", "field1": 23, "field2": 1325, "field3": 12436 },
                { "GT_RowId": 382, "title": "Test 382st row", "id": 382, "desc": "Lorem Ipsum is simply dummy 2673 text of the printing and typesetting", "info": "some info some info some info some info", "date": "21:56:58 10:05:2018", "field1": 97, "field2": 1616, "field3": 12727 }
                ]}
        />
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

it('renders Tools with empty Search/PagesSelector', () => {
    const tree = renderer.create(
        <Tools
            defaultPerPage={100}
            display="left"
            lang="en"
            perPage={50}
            perPageRows={[ 25, 50, 100, 200 ]}
            search=""
            selectedRows={[]}
            struct={{
                search: [ 'top', 'bottom' ],
                rowsSelector: [ 'desc', 'top', 'bottom' ],
                pagination: [ 'bottom' ],
                download: {
                    csv: true
                },
            }}
            tableOpts={{
                buttons: [
                    { // reload button to fetch content manually
                        extended: 'editor_reload',
                        editor: editor,
                        triggerAfter: (() => {
                            console.log('after reload')
                        }), triggerBefore: (() => {
                            console.log('before reload')
                        })
                    },
                    {// to set CSV Download button
                        extended: 'editor_csv',
                        editor: editor
                    },
                    {
                        extended: 'editor_create', editor: editor, triggerAfter: (function () {
                            console.log('after create')
                        }), triggerBefore: (function () {
                            console.log('before create')
                        })
                    },
                    {
                        extended: 'editor_edit', editor: editor, triggerBefore: (function () {
                            console.log('before edit')
                        })
                    },
                    {
                        extended: 'editor_remove', editor: editor, triggerAfter: (function () {
                            console.log('after del')
                        })
                    }
                ],
                buttonsPosition: [ 'top', 'bottom' ],
                theme: 'std'
            }}
            updatePerPage={(e) => {}}
            doSearch={(e) => {}}
        />
    ).toJSON()
    expect(tree).toMatchSnapshot()
})
