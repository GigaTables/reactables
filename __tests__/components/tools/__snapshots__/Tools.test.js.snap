exports[`test renders Tools correctly 1`] = `
<div
  className="gt_head_tools">
  <div
    className="gte_buttons_container"
    data-action="reload"
    onClick={[Function]}>
    <div
      className="gte_button"
      data-action="reload">
      <span
        data-action="reload">
        Reload
      </span>
    </div>
    <div
      className={undefined} />
  </div>
  <div
    className="gte_buttons_container"
    onClick={[Function]}>
    <div
      className="gte_button">
      <span>
        CSV Export
      </span>
    </div>
    <div
      className={undefined} />
  </div>
  <div
    className="gte_buttons_container"
    data-action="create"
    onClick={[Function]}>
    <button
      className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button">
      <span
        className="MuiButton-label-2">
        New
      </span>
      <span
        className="MuiTouchRipple-root-30" />
    </button>
  </div>
  <div
    className="gte_buttons_container"
    data-action="edit"
    onClick={undefined}>
    <button
      className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 MuiButton-disabled-21"
      disabled={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button">
      <span
        className="MuiButton-label-2">
        Edit
      </span>
    </button>
  </div>
  <div
    className="gte_buttons_container"
    data-action="delete"
    onClick={undefined}>
    <button
      className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 MuiButton-disabled-21"
      disabled={true}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button">
      <span
        className="MuiButton-label-2">
        Delete
      </span>
    </button>
  </div>
  <div
    className={undefined}>
    <span>
      Show
       
    </span>
    <span>
      <div
        className="MuiInputBase-root-57 MuiInput-root-44 MuiInput-underline-48"
        onClick={[Function]}>
        <div
          className="MuiSelect-root-37">
          <select
            aria-invalid={undefined}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiSelect-select-38 MuiInputBase-input-67 MuiInput-input-52"
            defaultValue={undefined}
            disabled={undefined}
            id="pages-selector"
            name="pages-selector"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type={undefined}
            value={50}>
            <option
              value={25}>
              25
            </option>
            <option
              value={50}>
              50
            </option>
            <option
              value={100}>
              100
            </option>
            <option
              value={200}>
              200
            </option>
          </select>
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-74 MuiSelect-icon-43"
            color={undefined}
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24">
            <path
              d="M7 10l5 5 5-5z" />
          </svg>
        </div>
      </div>
    </span>
    <span>
       
      entries
    </span>
  </div>
  <div
    className="">
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-83 MuiFormControl-marginDense-85">
      <label
        className="MuiFormLabel-root-94 MuiInputLabel-root-87 MuiInputLabel-formControl-88 MuiInputLabel-animated-91 MuiInputLabel-marginDense-89"
        data-shrink={false}
        htmlFor="standard-search">
        Search
      </label>
      <div
        className="MuiInputBase-root-57 MuiInput-root-44 MuiInput-underline-48 MuiInputBase-formControl-58 MuiInput-formControl-45 MuiInputBase-marginDense-64"
        onClick={[Function]}>
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInputBase-input-67 MuiInput-input-52 MuiInputBase-inputMarginDense-68 MuiInput-inputMarginDense-53"
          defaultValue={undefined}
          disabled={false}
          id="standard-search"
          name="search"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={false}
          rows={undefined}
          type="text"
          value="" />
      </div>
    </div>
    <div
      className={undefined}>
       
    </div>
  </div>
  <div
    className={undefined} />
</div>
`;

exports[`test renders Tools with empty Search/PagesSelector 1`] = `
<div
  className="gt_head_tools">
  
  
  <div
    className={undefined} />
</div>
`;
